getodk/collect
# Ensure that GPS is turned off on activity pause or exit.

Issues: Closes #3241.
Scope: MapFragments
Requested reviewers: @lognaturel 

#### User-visible changes

GPS location tracking is now properly turned off (the location icon disappears from the status bar) when a geo activity is paused or exited.  Previously, in some situations, the GPS would stay on indefinitely.

As an unfortunate side effect, the GPS lock is reset when the screen is rotated.  An existing GPS lock will be lost and the user will need to wait to re-acquire the lock (though usually this only takes a few seconds).  Previously, when the screen was rotated, the GPS lock state was preserved.

#### Internal changes

Previously, each Activity had to turn GPS on and off in its `onStart()` and `onStop()` methods.  Now, it's the `MapFragment`'s responsibility to remember what GPS state the activity wanted, and handle turning the GPS on and off in its own (Fragment) `onStart()` and `onStop()` methods.

Similarly, previously, each Activity had to handle a corner case where the `MapFragment` was detached before calling the `readyListener`.  Each `MapFragment` now avoids this corner case.

All the MapFragments now log a message when they start or stop getting location updates, so it's a little easier to confirm that this is working properly by filtering the logcat on the string `location update`.

#### Verification performed

For each of the cases listed below, I performed the following steps (looking at the GPS icon in the status bar to verify whether GPS is on):
1. Start the activity.  Verify that GPS is on.
2. Press Home.  Verify that GPS is off.
3. Return to Collect by tapping the app icon.  Verify that GPS is on.
4. Repeat 2 & 3 a few times.
5. Press the app switcher button.  Go to another app.  Verify that GPS is off.
6. Return to Collect using the app switcher.  Verify that GPS is on.
7. Repeat 5 & 6 a few times.
8. Rotate the screen several times, quickly.  Verify that GPS is still on.
9. Press Back to quit the activity and return to the form question.  Verify that GPS is off.

Here are all the cases tested:

Select basemap source "Google"
  - Start GeoPoint; perform verification.
  - Start GeoTrace; perform verification.
  - Start GeoShape; perform verification.

Select basemap source "OpenStreetMap"
  - Start GeoPoint; perform verification.
  - Start GeoTrace; perform verification.
  - Start GeoShape; perform verification.

Select basemap source "Mapbox"
  - Start GeoPoint; perform verification.
  - Start GeoTrace; perform verification.
  - Start GeoShape; perform verification.
